### PR TITLE
pcap miner 2

### PR DESCRIPTION
**Pull request by orlichter1**
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/18521

## Description
A new improved implementation of PCAP Miner.

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](https://github.com/demisto/content/edit/PCAP_miner2/Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml?pr=%2Fdemisto%2Fcontent%2Fpull%2F6228)

### CODE REVIEW

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  description: The emails extracted from the file.
+  type: String
+- contextPath: PcapResults.Regex
+  description: The regular expressions specified in `extract_regex`,extracted from
+    the file
+  type: String
+- contextPath: PcapResults.HTTP.ResponseStatusCode
+  description: The response code.
+  type: String
+- contextPath: PcapResults.HTTP.RequestVersion
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:159]: Change to 
PcapResultsHTTP.RequestVersion

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  description: The request version.
+  type: String
+- contextPath: PcapResults.HTTP.RequestCacheControl
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:162]: Change to
PcapResultsHTTP.RequestCacheControl

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  description: The chache control of the request.
+  type: String
+- contextPath: PcapResults.HTTP.ResponseDate
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:165]: Change to
PcapResultsHTTP.ResponseDate

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  description: The time and date of the response.
+  type: Date
+- contextPath: PcapResults.HTTP.RequestMethod
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:168]: Change to
PcapResultsHTTP.RequestMethod

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+  description: The request method.
+  type: String
+- contextPath: PcapResults.HTTP.RequestSourceIP
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:171]: Change to
PcapResultsHTTP.RequestSourceIP

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+  description: The source IP of the request.
+  type: String
+- contextPath: PcapResults.HTTP.ResponseContentType
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:174]: Change to
PcapResultsHTTP.ResponseContentType

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+  description: The response content type.
+  type: String
+- contextPath: PcapResults.HTTP.RequestAgent
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:177]: Change to
PcapResultsHTTP.RequestAgent

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+  description: The request agent.
+  type: String
+- contextPath: PcapResults.HTTP.RequestHost
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:180]: Change to
PcapResultsHTTP.RequestHost

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+  description: The request host.
+  type: String
+- contextPath: PcapResults.HTTP.ResponseVersion
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:183]: Change to
PcapResultsHTTP.ResponseVersion

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+  description: The response version.
+  type: String
+- contextPath: PcapResults.HTTP.ID
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:186]: Change to
PcapResultsHTTP.ID

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+  description: The ID of the http interaction.
+  type: String
+- contextPath: PcapResults.HTTP.RequestURI
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:189]: Change to
PcapResultsHTTP.RequestURI

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+  description: The request URI.
+  type: String
+- contextPath: PcapResults.HTTP.ResponseContentLength
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:192]: Change to
PcapResultsHTTP.ResponseContentLength

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+  description: The length of the rescponse content.
+  type: String
+- contextPath: PcapResults.HTTP.ResponseCodeDesc
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:195]: Change to
PcapResultsHTTP.ResponseCodecDesc

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+- contextPath: PcapResults.EndTime
+  description: The date and time of the last packet in the PCAP file.
+  type: String
+- contextPath: PcapResults.UniqueSourceIP
+  description: The number of unique IPs from which packets were transmitted.
+  type: Number
+- contextPath: PcapResults.UniqueDestIP
+  description: The number of unique IPs from to packets were transmitted.
+  type: Number
+- contextPath: PcapResults.Flow.Bytes
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:119]: Change to
PcapResultsFlow.Bytes

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+  description: The number of bytes transmitted in the flow.
+  type: String
+- contextPath: PcapResults.Flow.DestIP
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:122]: Change to
PcapResultsFlow.DestIP

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+  description: The destination IP of the flow.
+  type: String
+- contextPath: PcapResults.Flow.SourceIP
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:125]: Change to
PcapResultsFlow.SourceIP

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+  description: The source IP of the flow.
+  type: String
+- contextPath: PcapResults.Flow.SourcePort
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:128]: Change to
PcapResultsFlow.SourcePort

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+  description: The source port of the flow.
+  type: String
+- contextPath: PcapResults.Flow.DestPort
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:131]: Change to
PcapResultsFlow.DestPort

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+  description: The destination port of the flow.
+  type: String
+- contextPath: PcapResults.Flow.Duration
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:134]: Change to
PcapResultsFlow.Duration

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+  description: The duration of the flow (in seconds).
+  type: String
+- contextPath: PcapResults.Flow.EndTime
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:137]: Change to
PcapResultsFlow.EndTime

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+  description: The date time the flow ended.
+  type: Date
+- contextPath: PcapResults.Flow.StartTime
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:140]: Change to
PcapResultsFlow.StartTime

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+- contextPath: PcapResults.DNS.Response
+  description: The DNS response.
+  type: String
+- contextPath: PcapResults.DNS.Type
+  description: The type of the DNS request.
+  type: String
+- contextPath: PcapResults.Protocols
+  description: List of protocls found in the PCAP.
+  type: String
+script: '-'
```
Comment (David-BMS) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:213]: We are still missing other outputs besides DNS,HTTP for specific protocols

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py [hunk 23]
```diff
@@ -0,0 +1,693 @@
...
+            self.reg_message_code = re.compile(MESSAGE_CODE)
+        if 'FTP' in extracted_protocols:
+            self.reg_res_code = re.compile(RESPONSE_CODE)
+        if 'TELNET' in extracted_protocols:
+            self.telnet_data = set()  # type: set
+            self.telnet_commands = set()  # type: set
+        if homemade_regex:
+            self.reg_homemade = re.compile(self.homemade_regex)
+
+    def extract_context_from_packet(self, packet, layers_str: str, is_reg_extract: bool = False) -> None:
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py:101]: the function is way too long, please split

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py [hunk 24]
```diff
@@ -0,0 +1,693 @@
...
+
+            self.tcp_streams += 1
+            self.udp_streams += 1
+
+        except pyshark.capture.capture.TSharkCrashException:
+            raise ValueError("Could not find packets. Make sure that the file is a .cap/.pcap/.pcapng file "
+                             "or that filter is of the correct syntax.")
+
+        except Exception as error:
+            raise error
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py:484]: remove.
it's redundant and clears the exception's original traceback.

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py [hunk 25]
```diff
@@ -0,0 +1,693 @@
...
+                layers = strip(str(layers))
+                self.hierarchy[layers] = self.hierarchy.get(layers, 0) + 1  # type: ignore
+
+                # update times
+                packet_epoch_time = float(packet.frame_info.get('time_epoch'))
+                self.max_time = max(self.max_time, packet_epoch_time)
+                self.min_time = min(self.min_time, packet_epoch_time)
+
+                # count packets
+                self.num_of_packets += 1
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py:393]: can't this be outside the loop?
`self.num_of_packets = len(cap)`

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+  isArray: false
+  name: entry_id
+  required: true
+  secret: false
+- auto: PREDEFINED
+  default: false
+  description: What layers to ouput as context. Could be multiple layers seperated
+    by commas.
+  isArray: true
+  name: context_output
```
[the remark this answers is not in the document]
Comment (David-BMS) [reply Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.yml:13]: Pack name can be
Pcap Analysis

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py (hunk 23 above)
Comment (orlichter1) [reply Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py:101]: This function has a lot of ifs to handle all the different protocols. Should I add a different function for every protocol? I'm not sure if that would be more concise or faster. What do you think?

File: Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py (hunk 25 above)
Comment (orlichter1) [reply Packs/Legacy/Scripts/PcapMinerV2/PcapMinerV2.py:393]: Not really. As I don't want to load all the pcap to python's memory, the pyshark library doesn't keep the packets - everytime I call for a packet it goes and fetches it. This is what makes this script a bit slow but keeps the memory pretty clean. I think I found a way to load a bunch of packets at once so I'll dig a bit into that maybe to make this faster.

**Pull request closed without merging**